dbrp/service.go

```diff
@@ -0,0 +1,490 @@
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// Updates a mapping.
+// If another mapping with same organization ID, database, and retention policy exists, an error is returned.
+// Un-setting `Default` for a mapping will cause the first one to become the default.
+func (s *Service) Update(ctx context.Context, dbrp *influxdb.DBRPMappingV2) error {
+	if err := dbrp.Validate(); err != nil {
+		return ErrInvalidDBRP(err)
+	}
+	oldDBRP, err := s.FindByID(ctx, dbrp.OrganizationID, dbrp.ID)
+	if err != nil {
+		return ErrDBRPNotFound
+	}
+	// Overwrite fields that cannot change.
+	dbrp.ID = oldDBRP.ID
+	dbrp.OrganizationID = oldDBRP.OrganizationID
+	dbrp.BucketID = oldDBRP.BucketID
+	dbrp.Database = oldDBRP.Database
+
+	// If a dbrp with this orgID, db, and rp exists an error is returned.
+	if err := s.isDBRPUnique(ctx, *dbrp); err != nil {
+		return err
+	}
+
+	encodedID, err := dbrp.ID.Encode()
+	if err != nil {
+		return ErrInternalService(err)
+	}
+	b, err := json.Marshal(dbrp)
+	if err != nil {
+		return ErrInternalService(err)
+	}
+
+	return s.store.Update(ctx, func(tx kv.Tx) error {
+		bucket, err := tx.Bucket(bucket)
+		if err != nil {
+			return ErrInternalService(err)
+		}
+		if err := bucket.Put(encodedID, b); err != nil {
+			return err
+		}
+		compKey := indexForeignKey(*dbrp)
+		if dbrp.Default {
+			err = s.setAsDefault(tx, compKey, encodedID)
+		} else if oldDBRP.Default {
+			// This means default was unset.
+			// Need to find a new default.
+			first, ferr := s.getFirstBut(tx, compKey, encodedID)
+			if ferr != nil {
+				return ferr
+			}
+			if len(first) > 0 {
+				err = s.setAsDefault(tx, compKey, first)
+			}
+			// If no first was found, then this will remain the default.
+		}
+		return err
+	})
+}
+
+// Delete removes a mapping.
+// Deleting a mapping that does not exists is not an error.
+// Deleting the default mapping will cause the first one (if any) to become the default.
+func (s *Service) Delete(ctx context.Context, orgID, id influxdb.ID) error {
+	dbrp, err := s.FindByID(ctx, orgID, id)
+	if err != nil {
+		return nil
+	}
+	encodedID, err := id.Encode()
+	if err != nil {
+		return ErrInternalService(err)
+	}
+	return s.store.Update(ctx, func(tx kv.Tx) error {
+		bucket, err := tx.Bucket(bucket)
+		if err != nil {
+			return ErrInternalService(err)
+		}
+		compKey := indexForeignKey(*dbrp)
+		if err := bucket.Delete(encodedID); err != nil {
+			return err
+		}
+		if err := s.byOrgAndDatabase.Delete(tx, compKey, encodedID); err != nil {
+			return ErrInternalService(err)
+		}
+		// If this was the default, we need to set a new default.
+		var derr error
+		if dbrp.Default {
+			first, err := s.getFirstBut(tx, compKey, encodedID)
+			if err != nil {
+				return err
+			}
+			if len(first) > 0 {
+				derr = s.setAsDefault(tx, compKey, first)
+			} else {
+				// This means no other mapping is in the index.
+				// Unset the default
+				derr = s.unsetDefault(tx, compKey)
```

Comment:
Same goes here as for Update. Should we fail Delete if there is no suitable alternative `default`?

Reply:
I think this is ok as it is.

Because otherwise one could never delete mappings for a database